### PR TITLE
Modify RunResources usage

### PR DESCRIPTION
**Pull request by pefontana**
# Modify the usage of the RunResources struct

Reason for the changes: The fn execute_hint should receive a mutable reference of the RunResources, so blockifier and starknet_in_rust can count the step execution in recursive contracts calls.
CairoRunner::run_from_entrypoint should receive a  &mut RunResources as input instead of the `&mut Option<RunResources>`

Modify field type `RunResources.n_steps: Option<usize>,`

Changes in Apis:
*  CairoRunner::run_until_pc: Now receive a `&mut RunResources` instead of an `&mut Option<RunResources>`
*  CairoRunner::run_from_entrypoint: Now receive a `&mut RunResources` instead of an `&mut Option<RunResources>`
* VirtualMachine::Step: Add  &mut RunResources as input
* Trait HintProcessor::execute_hint: Add  &mut RunResources as an input 

Description of the pull request changes and motivation.

## Checklist
- [ ] Linked to Github Issue
- [ ] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [x] CHANGELOG has been updated.

### CODE REVIEW

File: vm/src/vm/runners/cairo_runner.rs [hunk 1]
```diff
@@ -78,24 +78,34 @@ impl From<Vec<MaybeRelocatable>> for CairoArg {
...
     n_steps: usize,
+    consume_steps: bool,
 }
 
 impl RunResources {
     pub fn new(n_steps: usize) -> Self {
-        RunResources { n_steps }
+        RunResources {
+            n_steps,
+            consume_steps: true,
```
Comment (fmoletta) [inline vm/src/vm/runners/cairo_runner.rs:88]: Does this need to be a field?
Can't it just be a public method that returns `n_steps.is_positive()`

File: vm/src/vm/runners/cairo_runner.rs [hunk 2]
```diff
@@ -78,24 +78,34 @@ impl From<Vec<MaybeRelocatable>> for CairoArg {
...
 
     pub fn consumed(&self) -> bool {
-        self.n_steps == 0
+        self.consume_steps && self.n_steps == 0
     }
 
     pub fn consume_steps(&mut self) {
-        self.n_steps -= 1;
+        if self.consume_steps {
+            self.n_steps -= 1;
```
Comment (fmoletta) [inline vm/src/vm/runners/cairo_runner.rs:98]:
```suggestion
            self.n_steps.saturating_sub(1);
```

File: vm/src/vm/runners/cairo_runner.rs [hunk 3]
```diff
@@ -78,24 +78,34 @@ impl From<Vec<MaybeRelocatable>> for CairoArg {
...
+            consume_steps: true,
+        }
     }
 
     pub fn consumed(&self) -> bool {
-        self.n_steps == 0
+        self.consume_steps && self.n_steps == 0
     }
 
     pub fn consume_steps(&mut self) {
```
Comment (fmoletta) [inline vm/src/vm/runners/cairo_runner.rs:96]:
```suggestion
    pub fn consume_step(&mut self) {
```

File: vm/src/vm/runners/cairo_runner.rs [hunk 4]
```diff
@@ -78,24 +78,34 @@ impl From<Vec<MaybeRelocatable>> for CairoArg {
 #[derive(Clone, Default, Debug, PartialEq)]
 pub struct RunResources {
     n_steps: usize,
+    consume_steps: bool,
```
Comment (fmoletta) [inline vm/src/vm/runners/cairo_runner.rs:81]:
```suggestion
    should_consume_steps: bool,
```
nit

File: vm/src/vm/runners/cairo_runner.rs [hunk 5]
```diff
@@ -78,24 +78,34 @@ impl From<Vec<MaybeRelocatable>> for CairoArg {
...
+        if self.consume_steps {
+            self.n_steps -= 1;
+        }
     }
 
     pub fn get_n_steps(&self) -> usize {
         self.n_steps
     }
+
+    pub fn get_consume_steps(&self) -> bool {
```
Comment (fmoletta) [inline vm/src/vm/runners/cairo_runner.rs:106]:
```suggestion
    pub fn should_consume_steps(&self) -> bool {
```

File: vm/src/vm/runners/cairo_runner.rs [hunk 6]
```diff
@@ -517,9 +527,9 @@ impl CairoRunner {
...
 
-    fn consumed(&self, run_resources: &mut Option<RunResources>) -> bool {
-        if let Some(r) = run_resources.as_ref() {
-            r.consumed()
+    fn consumed(&self, run_resources: &mut RunResources) -> bool {
+        if run_resources.consume_steps {
+            run_resources.consumed()
         } else {
             false
         }
```
Comment (fmoletta) [inline vm/src/vm/runners/cairo_runner.rs:535]: This method seems redundant given `RunResources::consumed`

**Review by Oppen — APPROVED** [review]

File: vm/src/vm/runners/cairo_runner.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pefontana) [reply vm/src/vm/runners/cairo_runner.rs:81]: good idea!

File: vm/src/vm/runners/cairo_runner.rs (hunk 2 above)
Comment (pefontana) [reply vm/src/vm/runners/cairo_runner.rs:98]: Good catch!

File: vm/src/vm/runners/cairo_runner.rs (hunk 6 above)
Comment (pefontana) [reply vm/src/vm/runners/cairo_runner.rs:535]: true

File: vm/src/vm/runners/cairo_runner.rs (hunk 5 above)
Comment (pefontana) [reply vm/src/vm/runners/cairo_runner.rs:106]: I end up with the n_steps: Option<usize> solution

File: vm/src/vm/runners/cairo_runner.rs (hunk 4 above)
Comment (pefontana) [reply vm/src/vm/runners/cairo_runner.rs:81]: I end up with the n_steps: Option<usize> solution

**Review by juanbono — APPROVED** [review]

File: vm/src/hint_processor/hint_processor_definition.rs [hunk 7]
```diff
@@ -27,6 +28,10 @@ pub trait HintProcessor {
         hint_data: &Box<dyn Any>,
         //Constant values extracted from the program specification.
         constants: &HashMap<String, Felt252>,
+        // RunResources keep track of the steps executed by the VirtualMachine.
```
Comment (fmoletta) [inline vm/src/hint_processor/hint_processor_definition.rs:31]:
```suggestion
        // RunResources keeps track of the steps executed by the VirtualMachine.
```

**Review by fmoletta — APPROVED** [review]

**Pull request merged**